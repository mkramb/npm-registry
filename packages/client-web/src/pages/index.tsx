import Head from 'next/head';

function HomePage() {
  return (
    <>
      <Head>
        <title>NPM registry</title>
      </Head>
      <div className="container">
        <h1 className="title">
          Welcome to <a href="https://nextjs.org">Next.js!</a>
        </h1>
      </div>
    </>
  );
}

export default HomePage;
