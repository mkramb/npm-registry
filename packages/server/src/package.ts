import got from 'got';
import * as async from 'async';
import { RequestHandler } from 'express';
import { maxSatisfying } from 'semver';
import { NPMPackage, RemotePackageRoot } from './types';

let queue: async.QueueObject<RemotePackageRoot>;
let dependencyTree: RemotePackageRoot;

const MAX_CONCURRENCY = process.env.MAX_CONCURRENCY ?? 10;

/**
 * Attempts to retrieve package data from the npm registry and return it
 */
export const getPackage: RequestHandler = async function (req, res, next) {
  const { name, version } = req.params;

  queue = async.queue<RemotePackageRoot>((task, done) => {
    getDependencies(task, done);
  }, MAX_CONCURRENCY);

  try {
    const npmPackage= await got(`https://registry.npmjs.org/${name}`).json<NPMPackage>();
    const dependencies = npmPackage?.versions?.[version]?.dependencies ?? {};

    dependencyTree = {
      name,
      version,
      dependencies: {},
    };

    for (const [ depName, depVersion ] of Object.entries(dependencies)) {
      queue.push({
        name: depName,
        version: depVersion,
        dependencies: dependencyTree.dependencies,
      });
    }

    if (queue.length() > 0) {
      await queue.drain();
    }

    return res
      .status(200)
      .json(dependencyTree);
  } catch (error) {
    return next(error);
  }
};

async function getDependencies(task: RemotePackageRoot, done: async.ErrorCallback<Error>) {
  const npmPackage = await got(`https://registry.npmjs.org/${task.name}`).json<NPMPackage>();
  const selectedVersion = maxSatisfying(Object.keys(npmPackage.versions), task.version);

  if (selectedVersion) {
    const newDependencies = npmPackage.versions[selectedVersion].dependencies ?? {};

    task.dependencies[task.name] = {
      version: selectedVersion,
      dependencies: {},
    };

    for (const [ name, version ] of Object.entries(newDependencies)) {
      queue.push({
        name,
        version,
        dependencies: task.dependencies[task.name].dependencies,
      });
    }
  }

  done();
}