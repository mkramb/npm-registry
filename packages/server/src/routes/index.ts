import express from 'express';

import { getPackage } from './getPackage';

function routes(router: express.Router): express.Router {
  return router.get('/package/:name/:version', getPackage);
}

export { routes };
