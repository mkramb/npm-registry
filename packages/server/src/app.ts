import http from 'http';
import cors from 'cors';
import helmet from 'helmet';
import logger from 'express-pino-logger';
import { createTerminus } from '@godaddy/terminus';
import { v4 as uuidv4 } from 'uuid';
import express from 'express';

import { notFound } from './middlewares/notFound';
import { routes } from './routes';

async function onHealthCheck() {
  // We can add any custom check
  // throw on error
  Promise.resolve();
}

/**
 * Bootstrap the application framework
 */
export function createApp(): http.Server {
  const app = express();
  const router = express.Router();

  const server = http.createServer(
    app
      .use(express.json())
      .use(helmet())
      .use(cors())
      .use(
        logger({
          genReqId: () => uuidv4(),
        })
      )
      .use(routes(router))
      .use(notFound)
  );

  createTerminus(server, {
    healthChecks: {
      '/healthcheck': onHealthCheck,
    },
  });

  return server;
}
