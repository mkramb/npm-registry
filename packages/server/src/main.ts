import { createApp } from './app';

const port = process.env.PORT ?? 4000;

async function main() {
  createApp().listen(port);
  console.info(`Server listening at http://localhost:${port}`);
}

main();
