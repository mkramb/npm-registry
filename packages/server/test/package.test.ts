import got from 'got';
import { Server } from 'http';
import { createApp } from '../src/app';
import { RemotePackageRoot } from '../src/types';

describe('/package/:name/:version endpoint', () => {
  let server: Server;
  let port: number;

  beforeAll(() => {
    return new Promise<void>((resolve, reject) => {
      server = createApp().listen(0, () => {
        const addr = server.address();

        if (addr && typeof addr === 'object') {
          port = addr.port;
          return resolve();
        }

        reject('Unexpected address ${addr} for server');
      });
    });
  });

  afterAll(() => {
    return new Promise<void>((resolve) => {
      server.close(() => resolve());
    });
  });

  it('react 16.13.0', async () => {
    const packageName = 'react';
    const packageVersion = '16.13.0';

    const request = got(`http://localhost:${port}/package/${packageName}/${packageVersion}`);
    const toJSON = request.json<RemotePackageRoot>();

    const [ response, json ] = await Promise.all([ request, toJSON ]);

    expect(response.statusCode).toEqual(200);
    expect(json.name).toEqual(packageName);
    expect(json.version).toEqual(packageVersion);
    expect(json.dependencies).toMatchSnapshot();
  });
});