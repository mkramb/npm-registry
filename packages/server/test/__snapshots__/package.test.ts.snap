// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/package/:name/:version endpoint react 16.13.0 1`] = `
Object {
  "loose-envify": Object {
    "dependencies": Object {
      "js-tokens": Object {
        "dependencies": Object {},
        "version": "4.0.0",
      },
    },
    "version": "1.4.0",
  },
  "object-assign": Object {
    "dependencies": Object {},
    "version": "4.1.1",
  },
  "prop-types": Object {
    "dependencies": Object {
      "loose-envify": Object {
        "dependencies": Object {
          "js-tokens": Object {
            "dependencies": Object {},
            "version": "4.0.0",
          },
        },
        "version": "1.4.0",
      },
      "object-assign": Object {
        "dependencies": Object {},
        "version": "4.1.1",
      },
      "react-is": Object {
        "dependencies": Object {},
        "version": "16.13.1",
      },
    },
    "version": "15.7.2",
  },
}
`;
